diff --git a/js/bootstrap.js b/js/bootstrap.js
index 86bccbd..c1a319b 100644
--- a/js/bootstrap.js
+++ b/js/bootstrap.js
@@ -125,6 +125,10 @@ var Drupal = Drupal || {};
     bootstrapAnchor: function (element) {
       element.validAnchor = element.nodeName === 'A' && (location.hostname === element.hostname || !element.hostname) && element.hash.replace(/#/,'').length;
       element.scrollTo = function(event) {
+        // Bail if element has no hash.
+        if (element.hash === '') {
+          return;
+        }
         var attr = 'id';
         var $target = $(element.hash);
         if (!$target.length) {
